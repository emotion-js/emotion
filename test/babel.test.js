/* eslint-disable jsx-quotes,no-useless-escape,no-template-curly-in-string */
/* eslint-env jest */
const babel = require('babel-core')
const plugin = require('../src/babel')

describe('emotion/babel', () => {
  describe('babel styled component', () => {
    test('no use', () => {
      const basic = 'styled.h1``'
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('no dynamic', () => {
      const basic = 'styled.h1`color:blue;`'
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('basic', () => {
      const basic = "const H1 = styled.h1`font-size: ${fontSize + 'px'};`"
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('function call', () => {
      const basic = "styled(MyComponent)`font-size: ${fontSize + 'px'};`"
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('basic fragment', () => {
      const basic = `
        const frag = fragment\`color: green\`;
        styled.h1\`font-size: 20px; @apply \${frag};\``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('fragment kitchen sink', () => {
      const basic = `
        const frag = fragment\`color: green; background-color: \${backgroundColor}\`;
        const frag1 = fragment\` width: 20px; name: some-frag-name; \`
        const frag2 = fragment\` height: 20px; @apply \${frag1}; \`
        styled.h1\`font-size: \${fontSize + 'px'}; name: some-name; @apply \${frag}; @apply \${frag2}\``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('css basic', () => {
      const basic = `
        css\`
        margin: 12px 48px;
        color: #ffffff;
        display: flex;
        flex: 1 0 auto;
        color: blue;
        width: \$\{widthVar\};
      \``
      const { code } = babel.transform(basic, {
        plugins: [[plugin]]
      })
      expect(code).toMatchSnapshot()
    })

    test('css kitchen sink', () => {
      const basic = `
        const cls = css\`font-size: 58pt;margin: $\{margin};\`
        const frag = fragment\`padding: 8px;\`
        const fragB = fragment\`height: $\{heightVar};@apply \$\{frag};\`
        const cls2 = css\`
        @apply $\{frag};
        @apply $\{fragB};
        margin: 12px 48px;
        color: #ffffff;
        \`
      `
      const { code } = babel.transform(basic, {
        plugins: [[plugin]]
      })
      expect(code).toMatchSnapshot()
    })

    test('name is correct with no identifier', () => {
      const basic = `
        css\`
        margin: 12px 48px;
        color: #ffffff;
        \`
      `
      const { code } = babel.transform(basic, {
        plugins: [[plugin]]
      })
      expect(code).toMatchSnapshot()
    })

    test('attr', () => {
      const basic = `styled('input')\`
       margin: attr(margin);
       color: #ffffff;
       height: \$\{props => props.height * props.scale\};
       width: attr(width);
       color: blue;
       display: \$\{flex\};
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('attr with value type', () => {
      const basic = `styled('input')\`
        margin: attr(margin px);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('attr with default value', () => {
      const basic = `styled('input')\`
        margin: attr(margin, 16);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('attr with value type and default value', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('match works on multiple', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
        color: blue;
        padding: attr(padding em, 16);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })

    test('attr kitchen sink', () => {
      const basic = `styled('input')\`
        margin: attr(margin px, 16);
        padding: attr(padding em, 16);
        font-size: attr(fontSize ch, 8);
        width: attr(width %, 95);
        height: attr(height vw, 90);
        display: attr(display, flex);
      \``
      const { code } = babel.transform(basic, {
        plugins: [plugin]
      })
      expect(code).toMatchSnapshot()
    })
  })
})
