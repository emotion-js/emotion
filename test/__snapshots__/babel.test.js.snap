// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emotion/babel babel styled component attr 1`] = `
"styled('input', 'css-14vghnb', [function getMargin(props) {
       return props.margin;
}, props => props.height * props.scale, function getWidth(props) {
       return props.width;
}, flex], function createEmotionStyledRules(x0, x1, x2, x3) {
       return [\`.css-14vghnb { margin: \${x0};
       color: #ffffff;
       height: \${x1};
       width: \${x2};
       color: blue;
       display: \${x3}; }\`];
});"
`;

exports[`emotion/babel babel styled component attr kitchen sink 1`] = `
"styled('input', 'css-1lqib20', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}, function getPadding(props) {
        return props.padding ? props.padding : ('16' + 'em');
}, function getWidth(props) {
        return props.width ? props.width : ('95' + '%');
}, function getHeight(props) {
        return props.height ? props.height : ('90' + 'vw');
}, function getDisplay(props) {
        return props.display ? props.display : ('flex' + '');
}], function createEmotionStyledRules(x0, x1, x2, x3, x4) {
        return [\`.css-1lqib20 { margin: \${x0};
        padding: \${x1};
        font-size: attr(fontSize ch, 8);
        width: \${x2};
        height: \${x3};
        display: \${x4}; }\`];
});"
`;

exports[`emotion/babel babel styled component attr with default value 1`] = `
"styled('input', 'css-1luxokk', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + '');
}], function createEmotionStyledRules(x0) {
        return [\`.css-1luxokk { margin: \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component attr with value type 1`] = `
"styled('input', 'css-1luxokk', [function getMargin(props) {
        return props.margin + 'px';
}], function createEmotionStyledRules(x0) {
        return [\`.css-1luxokk { margin: \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component attr with value type and default value 1`] = `
"styled('input', 'css-1luxokk', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}], function createEmotionStyledRules(x0) {
        return [\`.css-1luxokk { margin: \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component basic 1`] = `
"const H1 = styled('h1', 'css-H1-1t8i2zo', [fontSize + 'px'], function createEmotionStyledRules(x0) {
  return [\`.css-H1-1t8i2zo { font-size: \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component basic fragment 1`] = `
"
const frag = fragment(\\"frag-frag-upmkpf\\", [], function createEmotionFragment() {
        return [\`.frag-frag-upmkpf { color: green }\`];
});
styled(\\"h1\\", \\"css-u3lc8x\\", [frag], function createEmotionStyledRules(x0) {
        return [\`.css-u3lc8x { font-size: 20px; \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component css basic 1`] = `
"
css(\\"css-cu78iu\\", [widthVar], function createEmotionRules(x0) {
        return [\`.css-cu78iu { margin: 12px 48px;
        color: #ffffff;
        display: -webkit-box;
        display: -moz-box;
        display: -ms-flexbox;
        display: -webkit-flex;
        display: flex;
        -webkit-flex: 1 0 auto;
        -ms-flex: 1 0 auto;
        flex: 1 0 auto;
        color: blue;
        width: \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component css kitchen sink 1`] = `
"
const cls = css(\\"css-cls-h5t3cu\\", [margin], function createEmotionRules(x0) {
        return [\`.css-cls-h5t3cu { font-size: 58pt;margin: \${x0}; }\`];
});
const frag = fragment(\\"frag-frag-zzm2ov\\", [], function createEmotionFragment() {
        return [\`.frag-frag-zzm2ov { padding: 8px; }\`];
});
const fragB = fragment(\\"frag-fragB-16v04hq\\", [heightVar, frag], function createEmotionFragment(x0, x1) {
        return [\`.frag-fragB-16v04hq { height: \${x0};\${x1}; }\`];
});
const cls2 = css(\\"css-cls2-1b4c3x0\\", [frag, fragB], function createEmotionRules(x0, x1) {
        return [\`.css-cls2-1b4c3x0 { \${x0};
        \${x1};
        margin: 12px 48px;
        color: #ffffff; }\`];
});"
`;

exports[`emotion/babel babel styled component fragment kitchen sink 1`] = `
"
const frag = fragment('frag-frag-10oh1gr', [backgroundColor], function createEmotionFragment(x0) {
        return [\`.frag-frag-10oh1gr { color: green; background-color: \${x0} }\`];
});
const frag1 = fragment('frag-some-frag-name-1ymjbmm', [], function createEmotionFragment() {
        return [\`.frag-some-frag-name-1ymjbmm { width: 20px; name: some-frag-name; }\`];
});
const frag2 = fragment('frag-frag2-q1t5oy', [frag1], function createEmotionFragment(x0) {
        return [\`.frag-frag2-q1t5oy { height: 20px; \${x0}; }\`];
});
styled('h1', 'css-some-name-9k8moo', [fontSize + 'px', frag, frag2], function createEmotionStyledRules(x0, x1, x2) {
        return [\`.css-some-name-9k8moo { font-size: \${x0}; name: some-name; \${x1}; \${x2} }\`];
});"
`;

exports[`emotion/babel babel styled component function call 1`] = `
"styled(MyComponent, 'css-1t8i2zo', [fontSize + 'px'], function createEmotionStyledRules(x0) {
  return [\`.css-1t8i2zo { font-size: \${x0}; }\`];
});"
`;

exports[`emotion/babel babel styled component match works on multiple 1`] = `
"styled('input', 'css-1j6h68f', [function getMargin(props) {
        return props.margin ? props.margin : ('16' + 'px');
}, function getPadding(props) {
        return props.padding ? props.padding : ('16' + 'em');
}], function createEmotionStyledRules(x0, x1) {
        return [\`.css-1j6h68f { margin: \${x0};
        color: blue;
        padding: \${x1}; }\`];
});"
`;

exports[`emotion/babel babel styled component name is correct with no identifier 1`] = `
"
css(\\"css-1cppx8p\\", [], function createEmotionRules() {
        return [\`.css-1cppx8p { margin: 12px 48px;
        color: #ffffff; }\`];
});"
`;

exports[`emotion/babel babel styled component no dynamic 1`] = `
"styled(\\"h1\\", \\"css-14ksm7b\\", [], function createEmotionStyledRules() {
  return [\`.css-14ksm7b { color: blue; }\`];
});"
`;

exports[`emotion/babel babel styled component no use 1`] = `"\\"h1\\";"`;
